### PR TITLE
Fix wrong TCS spill table index calculation

### PR DESCRIPTION
**Pull request by crystalJinamd**
Root cause:
On GFX9+, LS and HS are merged into a single shader. For tessellation control shader, the user data count and spill table index should be same as vertex shader.

Solution:
Take vertex user data into consideration on GFX9+ and have both VS and TCS when calculate TCS user data count.

### CODE REVIEW

File: patch/llpcPatchEntryPointMutate.cpp [hunk 1]
```diff
@@ -516,12 +517,22 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
                 availUserDataCount -= 1;
             }
 
-            if (builtInUsage.vs.baseVertex || builtInUsage.vs.baseInstance)
+	    // on GFX9+, LS and HS are merged into a single shader.
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: // on GFX9+, LS and HS are merged into a single shader.
->
// NOTE:  On GFX9+, Vertex shader (LS) and tessellation control shader (HS) are merged into a single shader.

File: patch/llpcPatchEntryPointMutate.cpp [hunk 2]
```diff
@@ -516,12 +517,22 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
...
+	    // The user data count of tessellation control shader should be same as vertex shader.
+	    auto pCurResUsage = pResUsage;
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: pCurResUsage  -> pCurrResUsage

File: patch/llpcPatchEntryPointMutate.cpp [hunk 3]
```diff
@@ -781,23 +791,37 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
     switch (m_shaderStage)
     {
     case ShaderStageVertex:
+    case ShaderStageTessControl:
         {
+	    // on GFX9+, LS and HS are merged into a single shader.
+	    // The user data count of tessellation control shader should be same as vertex shader.
+            auto pCurIntfData = pIntfData;
+	    auto pCurResUsage = pResUsage;
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: Alignment is incorrect
pCurResUsage -> pCurrResUsage

File: patch/llpcPatchEntryPointMutate.cpp [hunk 4]
```diff
@@ -781,23 +791,37 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
     switch (m_shaderStage)
     {
     case ShaderStageVertex:
+    case ShaderStageTessControl:
         {
+	    // on GFX9+, LS and HS are merged into a single shader.
+	    // The user data count of tessellation control shader should be same as vertex shader.
+            auto pCurIntfData = pIntfData;
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: pCurIntfData -> pCurrIntfData

File: patch/llpcPatchEntryPointMutate.cpp [hunk 5]
```diff
@@ -781,23 +791,37 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
...
+	    auto pCurResUsage = pResUsage;
+
+            if ((m_pContext->GetGfxIpVersion().major >= 9) &&
+                 m_shaderStage == ShaderStageTessControl &&
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: m_shaderStage == ShaderStageTessControl
->
(m_shaderStage == ShaderStageTessControl)

File: patch/llpcPatchEntryPointMutate.cpp [hunk 6]
```diff
@@ -516,12 +517,22 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
                 availUserDataCount -= 1;
             }
 
-            if (builtInUsage.vs.baseVertex || builtInUsage.vs.baseInstance)
+	    // on GFX9+, LS and HS are merged into a single shader.
+	    // The user data count of tessellation control shader should be same as vertex shader.
+	    auto pCurResUsage = pResUsage;
+            if ((m_pContext->GetGfxIpVersion().major >= 9) &&
+		 m_shaderStage == ShaderStageTessControl &&
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: m_shaderStage == ShaderStageTessControl
->
(m_shaderStage == ShaderStageTessControl)

File: patch/llpcPatchEntryPointMutate.cpp [hunk 7]
```diff
@@ -781,23 +791,37 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
     switch (m_shaderStage)
     {
     case ShaderStageVertex:
+    case ShaderStageTessControl:
         {
+	    // on GFX9+, LS and HS are merged into a single shader.
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: // on GFX9+, LS and HS are merged into a single shader.
->
// NOTE: On GFX9+, Vertex shader (LS) and tessellation control shader (HS) are merged into a single shader.

**Comment by amdvlk-admin** [comment]
### Test summary for commit db3345ec0685053755641b7eb119bb94401ddfb1
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

File: patch/llpcPatchEntryPointMutate.cpp [hunk 8]
```diff
@@ -516,12 +517,22 @@ FunctionType* PatchEntryPointMutate::GenerateEntryPointType(
                 availUserDataCount -= 1;
             }
 
-            if (builtInUsage.vs.baseVertex || builtInUsage.vs.baseInstance)
+	    // NOTE: On GFX9+, Vertex shader (LS) and tessellation control shader (HS) are merged into a single shader.
+	    // The user data count of tessellation control shader should be same as vertex shader.
+	    auto pCurrResUsage = pResUsage;
+            if ((m_pContext->GetGfxIpVersion().major >= 9) &&
+		(m_shaderStage == ShaderStageTessControl) &&
```
Comment (amdrexu) [inline patch/llpcPatchEntryPointMutate.cpp]: Just a minor issue. Pay attention to the position of "("

**Comment by amdvlk-admin** [comment]
### Test summary for commit 654f5a49bd45211348b7daa9c4eef83a859b18af
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details><summary>Build Failed</summary><ul>

/jenkins/workspace/open-source/llpc-github-pr/driver_build/drivers/pal/src/util/lnx/lnxHashProvider.h:36:10: fatal error: openssl/sha.h: No such file or directory

</ul></details>

**Comment by amdrexu** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 654f5a49bd45211348b7daa9c4eef83a859b18af
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev 1fc1a7d4248b4749c3df21eb48f7ae97b6cddf74</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**